cloud-inquisitor/instrumentation/newrelic/logrus_hook.go

```diff
@@ -0,0 +1,186 @@
+	if err != nil {
+		return errors.New(fmt.Sprintf("New Relic Logrus Hook Error: Error creating new request - %v", err.Error()))
+	}
+	logRequest.Header.Set("Content-Type", "application/json")
+	logRequest.Header.Set("X-License-Key", nrh.newRelickHookOpts.License)
+
+	fmt.Printf("sending logrus hook json: %v\n", string(logBytes))
+	resp, err := nrh.httpClient.Do(logRequest)
+
+	fmt.Println("logrus hook response code: ", resp.StatusCode)
+	if body, bodyerr := ioutil.ReadAll(resp.Body); bodyerr == nil {
+		fmt.Println("logurs hook response body: ", string(body))
+	} else {
+		fmt.Println("error reading body: ", bodyerr.Error())
+	}
+	resp.Body.Close()
+
+	if err != nil {
+		return errors.New(fmt.Sprintf("New Relic Logrus Hook: recieved error posting log - %v", err.Error()))
+	}
+
+	return nil
+}
+
+type NewRelicHookFormatter struct{}
+
+// Format will format a logrus.Entry in the following way:
+//   timestamp will be in time.Unix() (seconds since epoch)
+//   JSON format will follow the New Relic Detailed JSON body message
+//     https://docs.newrelic.com/docs/logs/new-relic-logs/log-api/introduction-log-api#message-attribute-parsin
+//   Common fields include:
+//     - application_name: settings.name
+//     - log_level: settings.log_level
+//
+//   Per Log Fields include:
+//     - Workflow UUID
+//     - Session UUID
+//     - Fields included in the WithFields logger call
+func (nrhf *NewRelicHookFormatter) Format(entry *logrus.Entry, opts NewRelicHookOpts) ([]byte, error) {
+	//time in seconds
+	timestamps := entry.Time.Unix()
+
+	// json common entry
+	commonEntry := map[string]interface{}{}
+	commonEntry["timestamp"] = timestamps
+	commonAttributes := map[string]interface{}{}
+	commonAttributes["log_level"] = entry.Level
+	commonAttributes["application_name"] = opts.ApplicationName
+	commonEntry["attributes"] = commonAttributes
+
+	// json log entry
+	logEntry := map[string]interface{}{}
+	logEntry["timestamp"] = timestamps
+	logEntry["message"] = entry.Message
+
+	logAttributes := map[string]interface{}{}
+	for k, v := range entry.Data {
+		switch v := v.(type) {
+		case error:
+			logAttributes[removeWhiteSpace(k)] = v.Error()
+		default:
+			logAttributes[removeWhiteSpace(k)] = v
+		}
+	}
+
+	if entry.HasCaller() {
+		funcVal := entry.Caller.Function
+		fileVal := fmt.Sprintf("%s:%d", entry.Caller.File, entry.Caller.Line)
+		logAttributes["func"] = funcVal
+		logAttributes["file"] = fileVal
+	}
+
+	logEntry["attributes"] = logAttributes
+
+	jsonEntry := map[string]interface{}{
+		"common": commonEntry,
+		"logs": []map[string]interface{}{
+			logEntry,
+		},
+	}
+
+	jsonBytes, err := json.Marshal([]map[string]interface{}{jsonEntry})
+	if err != nil {
+		hookError := errors.New(fmt.Sprintf("New Relic Logrus Hook: recieved error from formatter - %v", err.Error()))
+		return jsonBytes, hookError
+	}
+	return jsonBytes, err
+}
+
+func removeWhiteSpace(whitespacestring string) string {
```

Comment:
This function appears to do more than remove white space.   Line 181 seems to add title casing to the input.

Reply:
renamed to `removeWhiteSpaceAndTitleCase`